src/models/hybrid.py

```diff
@@ -0,0 +1,289 @@
+import torch
+
+from torch import nn, Tensor
+from typing import Callable
+
+from transformers import PretrainedConfig
+from transformers.modeling_outputs import BaseModelOutput
+
+from transformers.models.mamba.modeling_mamba import (
+    MambaMixer,
+    MambaBlock,
+    MambaConfig,
+)
+
+from transformers.models.llama.modeling_llama import (
+    LlamaRMSNorm,
+    LlamaAttention,
+    LlamaRotaryEmbedding,
+    LlamaMLP,
+    LlamaDecoderLayer,
+    LlamaConfig,
+)
+
+from transformers.models.gpt2.modeling_gpt2 import (
+    GPT2Attention,
+    GPT2MLP,
+    GPT2Block,
+    GPT2Config,
+)
+
+from .transformer import BackboneModel
+
+
+SUPPORTED_BLOCKS = [
+    "residual",
+    "rms norm",
+    "layer norm",
+    "absolute positional embedding",
+
+    "llama attention",
+    "llama attention no rope",
+    "gpt2 attention",
+    "mamba mixer",
+
+    "llama mlp",
+    "gpt2 mlp",
+
+    "llama block",
+    "gpt2 block",
+    "mamba block",
+]
+
+class ResidualMarker(nn.Module):
+    """
+    This class is a dummy module to mark where 
+    residual connections should be made
+    """
+
+    def __init__(self, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+        self.net = nn.Identity()
+    
+    def forward(self, *args, **kwargs):
+        return self.net(*args, **kwargs)
+
+
+class AbsolutePositionalEmbedding(nn.Module):
+
+    def __init__(self, max_num_positions, hidden_dim):
+        super().__init__()
+        self.positions = torch.arange(0, max_num_positions, dtype=torch.int)
+        self.embed = nn.Embedding(max_num_positions, hidden_dim)
+
+    def forward(self, inp: Tensor) -> Tensor:
+        *_, seq_len, _ = inp.shape
+        pos = self.positions.to(device=inp.device)
+        embeddings = self.embed(pos)[:seq_len]
+        return inp + embeddings
+
+class RotaryEmbeddingStub(LlamaRotaryEmbedding):
+    def __init__(self, *args, enable: bool = False, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.enable = enable
+
+    def forward(self, x, position_ids):
+        cos, sin = super().forward(x, position_ids)
+
+        if self.enable:
+            return cos, sin
+        return torch.ones_like(cos), torch.zeros_like(sin)
+
+def _make_llama_attention_factory(config: PretrainedConfig, layer_idx: int, use_rope: bool) -> Callable[[], nn.Module]:
+    def llama_attn_factory():
+        attn_module = LlamaAttention(config=config, layer_idx=layer_idx) # pyright: ignore[reportArgumentType]
+        old_posemb = attn_module.rotary_emb
+
+        attn_module.rotary_emb = RotaryEmbeddingStub(
+            dim=old_posemb.dim,
+            max_position_embeddings=old_posemb.max_position_embeddings,
+            base=old_posemb.base,
+            scaling_factor=old_posemb.scaling_factor,
+            device=None, # TODO: replace with proper value
```

Comment:
what is the proper value for this?

Reply:
This is unclear -- the [transformers code](https://github.com/huggingface/transformers/blob/4b3eb19fa7f359d25f62ca9108479f71de912ebc/src/transformers/models/llama/modeling_llama.py#L295) doesn't set it to be different. Removing TODO.